src/runmode-dpdk.c

```diff
@@ -1335,6 +1337,85 @@ static int DeviceConfigure(DPDKIfaceConfig *iconf)
     SCReturnInt(0);
 }
 
+static int DPDKRunModeIsIPS(void)
+{
+    /* Find initial node */
+    const char dpdk_node_query[] = "dpdk.interfaces";
+    ConfNode *dpdk_node = ConfGetNode(dpdk_node_query);
+    if (dpdk_node == NULL) {
+        FatalError(SC_ERR_DPDK_CONF, "Unable to get %s configuration node", dpdk_node_query);
+    }
+
+    const char default_iface[] = "default";
+    ConfNode *if_default = ConfNodeLookupKeyValue(dpdk_node, "interface", default_iface);
+    int nlive = LiveGetDeviceCount();
+    bool has_ips = false;
+    bool has_ids = false;
+    for (int ldev = 0; ldev < nlive; ldev++) {
+        const char *live_dev = LiveGetDeviceName(ldev);
+        if (live_dev == NULL)
+            FatalError(
+                    SC_ERR_INVALID_VALUE, "Unable to get device id %d from LiveDevice list", ldev);
+
+        ConfNode *if_root = ConfFindDeviceConfig(dpdk_node, live_dev);
+        if (if_root == NULL) {
+            if (if_default == NULL)
+                FatalError(SC_ERR_INVALID_VALUE, "Unable to get %s or %s  interface", live_dev,
+                        default_iface);
+
+            if_root = if_default;
+        }
+
+        const char *copymodestr = NULL;
+        if (ConfGetChildValueWithDefault(if_root, if_default, "copy-mode", &copymodestr) == 1) {
+            if (strcmp(copymodestr, "ips") == 0) {
+                has_ips = true;
+            } else {
+                has_ids = true;
+            }
+        } else {
+            has_ids = true;
+        }
+    }
+
+    if (has_ids && has_ips) {
+        for (int ldev = 0; ldev < nlive; ldev++) {
```

Comment:
is there a point to this loop if we'll unconditionally `FatalError` right after it?

Reply:
I thought we can let the user know which interfaces are set to the IDS/TAP mode before we exit Suri

Reply:
could this be done during the initial parsing instead of during an additional loop?

Reply:
I think unless the user would specify the overall mode outright (e.g. a command line parameter or a global config entry), we need to do a two-way pass to determine what is the overall runmode of Suricata. If the first interface would be ids and then the other two interfaces would be paired in IPS mode then you would not be able to warn the user about the first interface in IDS mode. 
Minor optimization:
I could break from the first loop as soon as I would encounter both runmodes set at the same time (`has_ips && has_ids` condition) but I thought that would unnecessarily clutter the code and "performance" is not an issue here.

Reply:
Can't we do something like this?
```diff
diff --git a/src/runmode-dpdk.c b/src/runmode-dpdk.c
index a0d1d2871d..9afd188c99 100644
--- a/src/runmode-dpdk.c
+++ b/src/runmode-dpdk.c
@@ -577,6 +577,9 @@ static int ConfigSetCopyIface(DPDKIfaceConfig *iconf, const char *entry_str)
     SCReturnInt(0);
 }

+static DpdkCopyModeEnum g_copy_mode = DPDK_COPY_MODE_NONE;
+static bool g_copy_mode_set = false;
+
 static int ConfigSetCopyMode(DPDKIfaceConfig *iconf, const char *entry_str)
 {
     SCEnter();
@@ -604,6 +607,15 @@ static int ConfigSetCopyMode(DPDKIfaceConfig *iconf, const char *entry_str)
         iconf->copy_mode = DPDK_COPY_MODE_IPS;
     }

+    if (!g_copy_mode_set) {
+        g_copy_mode_set = true;
+        g_copy_mode = iconf->copy_mode;
+    } else {
+        if (g_copy_mode != iconf->copy_mode) {
+            FatalError(SC_ERR_DPDK_CONF, "can't mix copy modes");
+        }
+    }
+
     SCReturnInt(0);
 }
```

Reply:
Possibly yes - would maybe keep it in the "enable IPS function API" to follow the same convention as AFP or Netmap. 

I wrote the original PR so that **all** erroneous interfaces are printed out in a single Suricata run. This approach exits after the first error.

Reply:
I think printing out all incorrect interfaces would be more user-friendly although the logic is a bit more complex. 
But if you want to go with this solution I'll redo it.

Reply:
For DPDK I prefer a harsh approach, to keep complexity low and behavior clear. The AFP/Netmap changes are fine, we can do further cleanups there later.